python/hail/methods/statgen.py

```diff
@@ -1688,56 +1677,105 @@ def pc_relate(dataset, k, maf, block_size=512, min_kinship=-float("inf"), statis
+        Block size of block matrices used in the algorithm.
+        Default given by :meth:`.BlockMatrix.default_block_size`.
 
     Returns
     -------
     :class:`.Table`
-        A :class:`.Table` mapping pairs of samples to estimations of their
-        kinship and identity-by-descent zero, one, and two.
+        A :class:`.Table` mapping pairs of samples to their pair-wise statistics.
+    """
+    mt = require_biallelic(mt, 'pc_relate')
+    mt = require_unique_samples(mt, 'pc_relate')
+    require_col_key_str(mt, 'pc_relate')  # FIXME relax
+
+    assert len(mt.col_key) == 1
+    ck_name = list(mt.col_key)[0]
+    ck_type = mt.col_key[0].dtype
+    assert ck_type == hl.tstr
+
+    if k and not scores:
+        # column order always preserved?
+        _, scores_table, _ = hwe_normalized_pca(mt, k, compute_loadings=False, as_array=True)
+    elif scores and not k:
+        uid = Env.get_uid()
+        mt = mt.annotate_cols(**{uid: scores[mt.col_key].scores})  # FIXME: check that all cols are annotated
+        scores_table = mt.select_cols(*mt.col_key, scores=mt[uid]).cols()
+    elif k and scores:
+        raise ValueError("pc_relate: exactly one of 'k' and 'scores' must be set, found both")
+    else:
+        raise ValueError("pc_relate: exactly one of 'k' and 'scores' must be set, found neither")
 
-        The fields of the resulting :class:`.Table` entries are of types: `i`:
-        :py:data:`.tstr`, `j`: :py:data:`.tstr`, `kin`: :py:data:`.tfloat64`, `k2`:
-        :py:data:`.tfloat64`, `k1`: :py:data:`.tfloat64`, `k0`:
-        :py:data:`.tfloat64`. The table is keyed by `i` and `j`.
+    scores_local = scores_table.collect()
+    col_key_java = list(map(lambda x: ck_type._convert_to_j(x[ck_name]), scores_local))
+    pcs = list(map(lambda x: x.scores, scores_local))
+
+    mt = mt.select_entries(gt=mt.GT.n_alt_alleles())
+    mt = mt.select_rows(mean_gt=agg.mean(mt.gt))
+    mean_imputed_gt = hl.or_else(hl.float64(mt.gt), mt.mean_gt)
+
+    if not block_size:
+        block_size = BlockMatrix.default_block_size()
+    g = BlockMatrix.from_entry_expr(mean_imputed_gt,
+                                    block_size=block_size)
+
+    int_statistics = {"phi": 0, "phik2": 1, "phik2k0": 2, "all": 3}[statistics]
 
-    """
-    require_col_key_str(dataset, 'pc_relate')
-    dataset = require_biallelic(dataset, 'pc_relate')
-    intstatistics = {"phi": 0, "phik2": 1, "phik2k0": 2, "all": 3}[statistics]
-    _, scores, _ = hwe_normalized_pca(dataset, k, False, True)
     return Table(
         scala_object(Env.hail().methods, 'PCRelate')
-            .apply(dataset._jvds,
-                   k,
-                   scores._jt,
-                   maf,
-                   block_size,
-                   min_kinship,
-                   intstatistics))
+        .apply(Env.hc()._jhc,
+               g._jbm,
+               jarray(Env.jvm().java.lang.Object, col_key_java),  # FIXME this should be avoidable
```

Comment:
this is avoidable by using integer column indices and mapping back the true keys after.